nselib/msrpctypes.lua

```diff
@@ -4522,6 +4522,28 @@ function unmarshall_SERVICE_STATUS(data, pos)
   return pos, result
 end
 
+--- Unmarshalls a null-terminated Unicode string based upon a 32-bit offset (LPTSTR)
+-- @param arguments The data being processed
+-- @param pos  The current position within the data
+-- @return The new position
+-- @return The string with null removed
+function unmarshall_lptstr(arguments, startpos)
+
+  local offset = 5
+
+  -- Unpacks the string bacsed on its length i.e starting position and ending position.
+  --local str = string.unpack("<c" .. string.format("%d", (endpos - startpos)), arguments, startpos + offset)
+  local str = ""
+  local s = ""
+
+  while s ~= "\0\0" do
+    s = string.unpack("<c2", arguments, startpos + offset)
+    str = str .. s
```

Comment:
instead of using string.unpack, you can use string.sub to grab a 2-byte substring. Then just continue calculating the current position and return `string.sub(arguments, startpos, curpos)`. This avoids repeated string building via concatenation, which can be very bad for performance due to repeated reallocation of `str`.

Reply:
Your idea of using `string.sub` degrades the performance because when we know the end position there is no point in iterating all chars finding for NULL byte. I think its better to pass endpos argument to `unmarshall_lptstr` to improve the performance.

Your views?

Reply:
With the above changes you suggested I made a PR #991 but for the sake of this script I added a local function in msrpc.lua as `optimized_unmarshall_lptstr()` which accepts end position of the string as a parameter and this is a bit faster than the function in #991 .